Add DShot module

bkueng opened the pull request:
This brings #12634 from @Igor-Misic and adds a new output module that uses the low-level DShot interface. The PR includes the commits from https://github.com/PX4/Firmware/pull/12790.

The dshot module is implemented as drop-in replacement for fmu, which means it handles the same modes so that other functionality is not lost - the duplication is however not nice.

I've flown this on the Omnibus BetaFPV tiny whoop with DShot600.

DShot is enabled on OmnibusF4SD and Pixracer.
**Caution**: Before flying this on a Pixracer, we need to double-check that there are no DMA resource conflicts.

Missing/next steps:
- Support for F7 (including DMA-able memory and flushing caches). @Igor-Misic is that something you're still interested in/ already looking into?
- Enable on all other boards.
- More testing, also on boards with IO.
- DShot Telemetry

Igor-Misic commented:
@bkueng, first congrats on this amazing work you have done so far, and for the first DShot PX4 flight :)

I went through F7 DMA drivers (Nuttx DMA for F7) and I think this may work just as it is. F4 and F7 share the same registers for DMA. As you helped me on Slack to understand FMUv5 boards I added defines to skip that unsupported IO Processor (F1 MCUs) building Dshot for them. Firmware is building for F7 boards, and it should use their own drivers (nuttx/arch/arm/src/stm32f7/smt32_dma.h) for DShot drivers. 
Unfortunately, I don't have any F7 board to run and check it.

dagar commented:
> The dshot module is implemented as drop-in replacement for fmu, which means it handles the same modes so that other functionality is not lost - the duplication is however not nice.

One idea I was going to explore was continue splitting all the functionality into separate minimal modules and rely on the underlying channel allocations in io_timer to prevent conflicts. Pin ownership would then become first come, first served.

    drivers/px4fmu split into:
          -> drivers/pwm_out
          -> drivers/input_capture
          -> i2c and peripheral reset -> drivers/boardctl (or almost anywhere else...)

Instead of modes the fmu module (now simply pwm) would try to take the required pins when loading a mixer. If there's a conflict we can tell the user exactly what's happening (trying to load a mixer with 6 outputs and pins 5 & 6 already taken for gpio, etc) and either completely abort, or only use a subset.

### Pros
 - smaller easier to understand modules
 - new standalone `input_capture` module could be used for fmu-v5 CAP pins as well
 - pwm module could potentially then be consolidated with linux_pwm_out (each linux board implementing a tiny PWM hal)
 - get rid of awkward rcS pin configuration 
https://github.com/PX4/Firmware/blob/f0ee0b5d495224cdc517c216bd304094df9f0817/ROMFS/px4fmu_common/init.d/rcS#L354-L369

### Cons
 - likely harder to convey the overall pin configuration to a user for configuration in QGC than if we went for a single unified px4fmu
 - additional constraints on timer boundaries we might need to respect

hamishwillee commented:
@bkueng Can you please ping me when Dshot is in a documentable form - I presume this is not going into PX4 v1.10? 

In the docs we we explicitly state that we don't support it in the docs - so we would need to remove that. If there is any explicit config or different hardware set up we'd want to capture that too.

bkueng commented:
> @bkueng, first congrats on this amazing work you have done so far, and for the first DShot PX4 flight :)

Thanks! The larger portion of the credits goes to you, without the lower level parts it would not be possible.

> I went through F7 DMA drivers (Nuttx DMA for F7) and I think this may work just as it is.

Indeed, it does! I expected more work would be needed.

---

@dagar yes that might work.

---

@hamishwillee sure. The short answer: as soon as this goes in it's usable (we need to enable it on all boards though).

### Updates
- added F7, enabled KakuteF7
- added (indirect) telemetry via UART and publish `esc_status`
- added CLI commands, for example:
  - let the ESC beep: `dshot beep1 -m 1`
  - permanently reverse motor direction (no more ESC resoldering :)
  ```
  dshot reverse -m 1
  dshot save -m 1
  ```
  - retrieve ESC infos: `dshot esc_info -m 2`:
```
nsh> dshot esc_info -m 2
INFO  [dshot] ESC Type: #TEKKO32_4in1#
INFO  [dshot] MCU Serial Number: 480022-001751-564e34-323420
INFO  [dshot] Firmware version: 32.60
INFO  [dshot] Rotation Direction: normal
INFO  [dshot] 3D Mode: off
INFO  [dshot] Low voltage Limit: off
INFO  [dshot] Current Limit: off
INFO  [dshot] LED 0: unsupported
INFO  [dshot] LED 1: unsupported
INFO  [dshot] LED 2: unsupported
INFO  [dshot] LED 3: unsupported
```

### Known Issues
Problems I came across while testing:
- Sometimes (quite rare) upon disarming the ESCs re-initialize, so there is something about the signal they don't like. I see this on 2 completely different quads (F7 vs F4 and different ESCs), so it's unlikely a HW-problem.
- Telemetry and sending of commands is not 100% reliable. Sometimes a command is not executed, and telemetry requests regularly and randomly time out. A cross-test with betaflight sending commands on the same HW does not show that issue.

hamishwillee commented:
> @hamishwillee sure. The short answer: as soon as this goes in it's usable (we need to enable it on all boards though).

And the long answer? :-) Because I take that to mean that these instructions will work without any change http://docs.px4.io/master/en/peripherals/pwm_escs_and_servo.html

bkueng commented:
This can be tested on Pixracer now, e.g. by setting `DSHOT_CONFIG` to 1200.

mhkabir commented:
@bkueng How can we get this working on Pixhawk 4?

hamishwillee commented on src/drivers/dshot/module.yaml:
```diff
@@ -0,0 +1,54 @@
+            description:
+                short: Configure DShot
+                long: |
+                    This enabled/disabled DShot. The different modes define different
```
enables/disables (i.e. use correct tense)

bkueng replied:
Such an obvious mistake... and I was even consistent.

hamishwillee commented on src/drivers/dshot/module.yaml:
```diff
@@ -0,0 +1,54 @@
+                    This enabled/disabled DShot. The different modes define different
+                    speeds, for example DShot150 = 150kb/s. Not all ESCs support all modes.
+                    
+                    Note: this enabled DShot on the FMU outputs. For boards with an IO it is the
```
enables

bkueng commented:
> @bkueng How can we get this working on Pixhawk 4?

I enabled it on v5, but only for the first 4 outputs (the next ones conflict with px4io
serial DMA (UART8_RX)). I had to disable RX DMA on the GPS port - can you test GPS as well?

mhkabir commented:
OK, let me have a look , thanks :)

Igor-Misic commented on platforms/nuttx/src/px4/stm/stm32_common/dshot/dshot.c:
```diff
@@ -0,0 +1,440 @@
+			first_motor += motors_number;
+
+			uint32_t dshot_data_size = motors_number * ONE_MOTOR_BUFF_SIZE;
+			rLIFCR(timer) |= dma_int_streamx_mask; //clear interrupt flags
```
From stream4 to stream7  rHIFCR should be cleared.
![image](https://user-images.githubusercontent.com/10188706/65550135-33599580-df1f-11e9-9e47-d52bb15b4d49.png)

bkueng replied:
Thanks, updated.

bkueng commented:
I've flown this quite a bit on Pixracer and KakuteF7 (Kopis 2 frame with telemetry), so that I'm confident it's working.

### Flight Performance
On both vehicles I've flown (250 size), I noticed considerably smoother flight performance - both audibly and visibly. Prop-wash handling is noticeably improved.
Part of this is certainly due to the 1kHz rate loop, and I have not (yet) checked how much of it is to attribute to DShot.

### Boards
DShot is enabled on these boards:
- Pixhawk (v3): enabled on all outputs, disabled RX DMA on TEL4 and IO debug port
- Pixhawk 4 (v5): enabled only on first 4 outputs, disabled RX DMA on GPS port
- Pixracer (v4): enabled on all outputs, disabled RX DMA on FrSky telem port
- OmnibusF4SD: enabled on all outputs, disabled RX DMA on USART1
- KakuteF7: enabled on all outputs, disabled RX DMA on UART4 (GPS port)

### Issues
Issues from above:
> Sometimes (quite rare) upon disarming the ESCs re-initialize, so there is something about the signal they don't like. I see this on 2 completely different quads (F7 vs F4 and different ESCs), so it's unlikely a HW-problem.

The problem is the flash-based param saving: in some cases the entire CPU stalls for ~300ms, which causes DShot not to update anymore (on PWM and Oneshot, the timers continue to trigger). If in-air, the motors might stop, so the param auto-saving is now deferred until disarming for flash-based param storage.

> Telemetry and sending of commands is not 100% reliable. Sometimes a command is not executed, and telemetry requests regularly and randomly time out. A cross-test with betaflight sending commands on the same HW does not show that issue.

Fixed by updating the timer config before triggering the DMA update.

### Logs
With a Pixracer:
- https://logs.px4.io/plot_app?log=eda0d033-187d-4f52-8bf3-8cc3c320ef71
- https://logs.px4.io/plot_app?log=cf6b496e-dbd2-4432-85ca-e558045803c9
- https://logs.px4.io/plot_app?log=48b1860c-d366-4b55-9af4-624fceb5b1cf

### Other
- CPU load is about 1% higher and ~1KB more RAM is used with DShot vs PWM (RAM is due to the DMA buffers)
- If for some reason the DShot signal stops updating while motors are spinning (e.g. actuator topic is not updated), the motors slowly stop and disarm.
- There is a pending refactoring from @Igor-Misic to use NuttX DMA methods, but it still has issues, so we can merge that separately.
- Once we use ITCM RAM on F7, we'll need to use DMA memory allocators for the DShot buffers.

@PX4/testflights can you test this on Pixhawk (v3), Pixracer and Pixhawk 4 on vehicles with ESCs that support DShot? You can follow the instructions here: https://github.com/PX4/px4_user_guide/pull/578/files#diff-58ea316b3b439c16b17ab4f924adc37a. Be careful, the vehicle might need some re-tuning.

I'm not sure if we want to pull this into v1.10?

dagar commented:
> Once we use ITCM RAM on F7, we'll need to use DMA memory allocators for the DShot buffers.

We're currently using non-DMA capable memory in the heap on F7. 

https://github.com/PX4/Firmware/blob/4f71c4e123a627ce0724b8065089c8e86038aebc/boards/px4/fmu-v5/nuttx-config/scripts/script.ld#L76

> On both vehicles I've flown (250 size), I noticed considerably smoother flight performance - both audibly and visibly. Prop-wash handling is noticeably improved.
> Part of this is certainly due to the 1kHz rate loop, and I have not (yet) checked how much of it is to attribute to DShot.

It would be interesting to compare a few flights at different loop rates. You can do this now with the `IMU_GYRO_RATEMAX` parameter. With a bit more work (drivers cleanup + drdy, standalone rate controller, optimization in a few key places) we should be able to get to 2 kHz "soon".
Related - https://github.com/betaflight/betaflight/issues/7327#issuecomment-452994940

Tony3dr commented:
@bkueng, unfortunately, we don't have vehicles with ESCs that support DShot. All our current setups have the Hobbywing XRotor 20A APAC Brushless ESC. Any that you recommend? Thanks!

mhkabir commented:
@PX4/testflights can you try this on a regular Pixhawk 4 vehicle with GPS? (no dshot)

mhkabir commented:
As for recommended ESCs, I've had good experiences with these:
- TMotorF55A Pro 55A 3-6s BLHeli32 4-in-1 ESC
- Holybro Tekko32 35A 4-in-1 BLHeli32 3-6s ESC.

Any recent BLHeli32 ESC should work though.

bkueng commented:
@Tony3dr  in addition to @mhkabir's recommendation: we generally use the 'Hobbywing XRotor 30A Micro' on F450-sized vehicles.

>  We're currently using non-DMA capable memory in the heap on F7.

Oh indeed. I switched back to a static buffer. Drawbacks are a couple of 100 bytes more memory used and the buffers are used even if DShot is not enabled (~1KB).

>  It would be interesting to compare a few flights at different loop rates.

I compared against `IMU_GYRO_RATEMAX=250` and `IMU_GYRO_RATEMAX=500`, all on DShot. Both increments (250 -> 500 and 500 -> 1000) make a noticeable difference in terms of smoothness and allow for higher rate P gains. It's likely that 2kHz brings additional improvements, albeit with diminishing returns. Would be interesting to test as well.
(with 250 the ESC's initialized twice, but then worked as usual)

bkueng commented:
> What's the worst case on v5 for comparison? Trimble GPS?

Yes. And I guess that leaves the testing with me. It still works, but idle time noticeably goes down, roughly 3.5%. So, not great.

dagar commented:
> > What's the worst case on v5 for comparison? Trimble GPS?
> 
> Yes. And I guess that leaves the testing with me. It still works, but idle time noticeably goes down, roughly 3.5%. So, not great.

How are things overall on that setup? If there's enough margin to tolerate the increase now I think there's plenty of opportunity to more than make up for it in the near future between DMA on SPI and and serial TX. I'm guessing SPI and mavlink tx currently account for more than half the cpu time.

bkueng commented:
I think it will work. I see that we're at about 75-80% both on a VTOL and MC with log streaming (and dual GPS on the MC).

Igor-Misic commented:
Guys, if you still didn't do a test flight with these changes I encourage you to do so. It works much smoother than before. :)

jorge789 commented:
**Tested on PixRacer V4:**
**Modes Tested**

Position Mode: Good.
Altitude Mode: Good.
Stabilized Mode: Good.
Mission Plan Mode (Automated): Good.
RTL: Good.

**- Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoin activate RTL and see landing behaviour.

**Notes:**
No issues noted, good flight in general.

**Log:**
https://review.px4.io/plot_app?log=930e9841-e3f9-48b2-8828-4e987e7666b8

https://review.px4.io/plot_app?log=caf2431b-4a7c-4a9f-9154-b75b1227e7b5

Junkim3DR commented:
### Tested on Pixhawk 4mini v5:

**Modes Tested**
- Position Mode: Good.
- Altitude Mode: Good.
- Stabilized Mode: Good.
- Mission Plan Mode (Automated): Good.
- RTL (Return To Land): Good.

**Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint then trigger RTL and see landing behavior.

**Notes**
No issue noted, good flight in general.

**Log:**
https://review.px4.io/plot_app?log=de4b7d15-2a26-4b43-9c28-7b6a55cd21e2

mrpollo commented:
This came up during the dev call October 9th, @dagar @julianoes can you guys help with a review?

hamishwillee commented:
@bkueng Is this going into v1.10? If not, then we should defer merging the dshot docs until after the 1.10 docs tree is created.

julianoes commented on msg/esc_status.msg:
```diff
@@ -1,16 +1,12 @@
 uint8 ESC_CONNECTION_TYPE_PPM = 0			# Traditional PPM ESC
 uint8 ESC_CONNECTION_TYPE_SERIAL = 1			# Serial Bus connected ESC
-uint8 ESC_CONNECTION_TYPE_ONESHOOT = 2			# One Shoot PPM
+uint8 ESC_CONNECTION_TYPE_ONESHOT = 2			# One Shot PPM
```
Shoot! :smile:

julianoes commented on src/drivers/dshot/dshot.cpp:
```diff
@@ -0,0 +1,1679 @@
+using namespace time_literals;
+
+/** Mode given via CLI */
+enum PortMode {
```
enum class would be nice.

julianoes commented on src/drivers/dshot/dshot.cpp:
```diff
@@ -0,0 +1,1679 @@
+class DShotOutput : public cdev::CDev, public ModuleBase<DShotOutput>, public OutputModuleInterface
+{
+public:
+	enum Mode {
```
Same here. And is this needed?

bkueng replied:
These are the things that come from `fmu`. We can possibly simplify, but preferably as a separate step.

julianoes replied:
Ok, maybe we can extract it from fmu and dshot later and improve it.

julianoes commented on src/drivers/dshot/dshot.cpp:
```diff
@@ -0,0 +1,1679 @@
+
+	// wait until main thread processed it
+	while (_new_command.load()) {
+		usleep(1000);
```
```suggestion
		px4_usleep(1000);
```

bkueng replied:
Right, updated.

julianoes commented on src/drivers/dshot/dshot.cpp:
```diff
@@ -0,0 +1,1679 @@
+	int max_time = 1000;
+
+	while (_request_esc_info.load() != nullptr && max_time-- > 0) {
+		usleep(1000);
```
```suggestion
		px4_usleep(1000);
```

julianoes commented on src/drivers/dshot/dshot.cpp:
```diff
@@ -0,0 +1,1679 @@
+	}
+
+
+	/* start the FMU if not running */
```
Remove comment.

bkueng replied:
Done

bkueng commented on boards/px4/fmu-v5/src/board_config.h:
```diff
@@ -671,6 +671,10 @@
 
+#define BOARD_NUM_IO_TIMERS 5
+
+#define BOARD_DSHOT_MOTOR_ASSIGNMENT {3, 2, 1, 0, 4, 5, 6, 7};
```
It's not needed for every board, and it is board-specific. It comes from the fact that DMA has certain ordering requirements (which are different from PWM).

julianoes replied:
Ah ok!

bkueng commented on boards/px4/fmu-v5/nuttx-config/stackcheck/defconfig:
```diff
@@ -218,7 +218,6 @@ CONFIG_UART8_RXDMA=y
 CONFIG_UART8_TXBUFSIZE=1500
 CONFIG_USART1_BAUD=57600
 CONFIG_USART1_RXBUFSIZE=600
-CONFIG_USART1_RXDMA=y
```
Yes: https://github.com/PX4/Firmware/pull/12854#issuecomment-536943871

bkueng commented:
Thanks for the reviews and tests everyone.
@hamishwillee looks like we can get it into v1.10.

MaEtUgR commented:
This is so awesome!!! :rocket:

I come from flying the Holybro Kopis 2 with the new DSHOT support and it flies that good now! It's really confidence-inspiring.
Next time I'll bring a GoPro but here's the eye cancer stream such that you get an impression:
https://youtu.be/XDT9YaBVlEs
Here are the two flight logs:
https://logs.px4.io/plot_app?log=786298ac-1b19-4019-be56-130c64a1baf0
https://logs.px4.io/plot_app?log=4c7c0414-e84e-47d6-9116-6fda9c0cb67e

Thanks a lot to @bkueng @dagar @Igor-Misic and everyone else who helped for doing all these low-level improvements that will drastically improve any low latency actuated vehicle's flight performance!

thomasderflieger commented:
Since being spoiled from flying racing-quadrocopters with DSHOT and telemetry for quite some time i am very happy to see those features being implemented into PX4. Thank you!

I'm currently building a VTOL with the CUAV V5 Nano Flight Controller. It has 3 Motors and 5 Servos, wich is also exactly the number of PWM outputs of the FC.  It is similar to the Convergence-airframe, but i can also tilt the rear Motor. 


> I enabled it on v5, but only for the first 4 outputs (the next ones conflict with px4io
> serial DMA (UART8_RX)). I had to disable RX DMA on the GPS port - can you test GPS as well?

Doas that mean if i enable DSHOT the first 4 outputs are going to use DSHOT or doas it depend on the motor-count.  If it is enabled on the first 4 outputs, i could not use nr. 4, but then im missing an PWM output for one of my Servos. 

The second question is regarding logging. Is it possible to log the ESC telemetry data and add it to the logfiles? That could help with analysis a lot.

hamishwillee commented:
@thomasderflieger Docs are ere: http://docs.px4.io/master/en/peripherals/dshot.html. You wire according to your mixer file/airframe reference: http://docs.px4.io/master/en/airframes/airframe_reference.html

@bkueng Would need to comment on logging.

bkueng commented:
> Doas that mean if i enable DSHOT the first 4 outputs are going to use DSHOT or doas it depend on the motor-count. If it is enabled on the first 4 outputs, i could not use nr. 4, but then im missing an PWM output for one of my Servos.

It's indepenent on the motor count and the current implementation does not allow for mixing DShot with PWM outputs on the FMU pins (on a board with an IO you could use PWM on the IO pins and DShot on the FMU pins).

> The second question is regarding logging. Is it possible to log the ESC telemetry data and add it to the logfiles? That could help with analysis a lot.

It already is. It is the `esc_status` message.

thomasderflieger commented:
> It's indepenent on the motor count and the current implementation does not allow for mixing DShot with PWM outputs on the FMU pins (on a board with an IO you could use PWM on the IO pins and DShot on the FMU pins).

Thats a pitty, was looking forward to using dshot for my VTOL, my current flight controller just has main-ports though. I assume that the Telemetry wont work without using dshot either, since the ESC's need a "telemetry-request message"?

> It already is. It is the esc_status message.

Great, thanks!

bkueng commented:
> Thats a pitty, was looking forward to using dshot for my VTOL, my current flight controller just has main-ports though. I assume that the Telemetry wont work without using dshot either, since the ESC's need a "telemetry-request message"?

Exactly. Yeah, that's a drawback, but with all configuration options and timer restrictions, it would become really complex.

ndepal commented:
> It's indepenent on the motor count and the current implementation does not allow for mixing DShot with PWM outputs on the FMU pins (on a board with an IO you could use PWM on the IO pins and DShot on the FMU pins).

Just do be sure, does that mean that it's not possible to use camera triggering (GPIO) on a PixRacer together with DShot?

bkueng commented:
> Just do be sure, does that mean that it's not possible to use camera triggering (GPIO) on a PixRacer together with DShot?

That is possible. Do you have a setup where you can test it?

ndepal commented:
yes we could probably test that pretty easily

ndepal commented on src/drivers/dshot/dshot.cpp:
```diff
@@ -0,0 +1,1678 @@
+	}
+
+	// publish when motor index wraps (which is robust against motor timeouts)
+	if (motor_index < _telemetry->last_motor_index) {
```
When we were testing dshot with just one ESC connected we could not get telemety info on `esc_status` because of this line. I don't know if 1 ESC is an actual use case (planes?), but if it is, would the following be better?

```
if (motor_index <= _telemetry->last_motor_index) {
```

bkueng replied:
Good point. I don't see any issues with that. Can you do a PR?

ndepal commented:
> That is possible. Do you have a setup where you can test it?

A quick bench test with camera triggering configured on pins 56 on a PixRacer (GPIO, active low) appears to work just fine.

bkueng commented:
Thanks for confirming
